vpr/SRC/route/route_tree_timing.c

```diff
@@ -479,16 +556,587 @@ void free_route_tree(t_rt_node * rt_node) {
+		if (g_rr_switch_inf[iswitch].buffered) {
+			if (root->R_upstream != rr_node[inode].R + g_rr_switch_inf[iswitch].R) {
+				vpr_printf_info("%d mismatch R upstream %e supposed %e\n", inode, root->R_upstream, 
+					rr_node[inode].R + g_rr_switch_inf[iswitch].R);
+				return false;
+			}
+		}
+		else if (root->R_upstream != rr_node[inode].R + root->parent_node->R_upstream + g_rr_switch_inf[iswitch].R) {
+			vpr_printf_info("%d mismatch R upstream %e supposed %e\n", inode, root->R_upstream, 
+				rr_node[inode].R + root->parent_node->R_upstream + g_rr_switch_inf[iswitch].R);
+			return false;
+		}
+	}
+	else if (root->R_upstream != rr_node[inode].R) {
+		vpr_printf_info("%d mismatch R upstream %e supposed %e\n", inode, root->R_upstream, rr_node[inode].R);
+		return false;
+	}
+
+	// check downstream C
+	t_linked_rt_edge* edge = root->u.child_list;
+	float C_downstream_children {0};
+	// sink, must not be congested
+	if (!edge) {
+		int occ = rr_node[inode].get_occ();
+		int capacity = rr_node[inode].get_capacity();
+		if (occ > capacity) {
+			vpr_printf_info("SINK %d occ %d > cap %d\n", inode, occ, capacity);
+			return false;
+		}
+	}
+	while (edge) {
+		if (edge->child->parent_node != root) {
+			vpr_printf_info("parent-child relationship not mutually acknowledged by parent %d->%d child %d<-%d\n", 
+				inode, edge->child->inode,
+				edge->child->inode, edge->child->parent_node->inode);
+			return false;
+		}
+		if (edge->iswitch != edge->child->parent_switch) {
+			vpr_printf_info("parent(%d)-child(%d) connected switch not equivalent parent %d child %d\n", 
+				inode, edge->child->inode, edge->iswitch, edge->child->parent_switch);
+			return false;
+		}
+
+		if (!g_rr_switch_inf[edge->iswitch].buffered)
+			C_downstream_children += edge->child->C_downstream;
+
+		if (!is_valid_route_tree(edge->child)) {
+			vpr_printf_info("subtree %d invalid, propagating up\n", edge->child->inode);
+			return false;
+		}
+		edge = edge->next;
+	}
+
+	if (root->C_downstream != C_downstream_children + rr_node[inode].C) {
+		vpr_printf_info("mismatch C downstream %e supposed %e\n", root->C_downstream, C_downstream_children + rr_node[inode].C);
+		return false;
+	}
+
+	return true;
+}
+
+bool is_uncongested_route_tree(t_rt_node* root) {
+	// make sure the nodes are legally connected
+	t_linked_rt_edge* edge {root->u.child_list};
+	
+	for (;;) {
+		int inode = root->inode;
+		if (rr_node[inode].get_occ() > rr_node[inode].get_capacity()) {
+			// vpr_printf_info("node %d occ %d > cap %d\n", inode, rr_node[inode].get_occ(), rr_node[inode].get_capacity());
+			return false;
+		}
+
+		// reached a sink
+		if (!edge) {
+			return true;
+		}
+		// join point (sibling edges)
+		else if (edge->next) {
+			// recursively update for each of its sibling branches
+			do {
+				if (!is_uncongested_route_tree(edge->child)) return false;
+				edge = edge->next;
+			} while (edge);
+			return true;
+		}
+
+		root = edge->child;
+		edge = root->u.child_list;
+	}
+}
```

Comment:
Should briefly comment the purpose of all the routines above.
Are some of the legality checks always on (e.g. check for a valid route tree on the final routing for each net)?